servicetalk-http-api/src/main/java/io/servicetalk/http/api/StreamingHttpServiceToOffloadedStreamingHttpService.java

```diff
@@ -70,12 +70,18 @@ public Single<StreamingHttpResponse> handle(final HttpServiceContext ctx,
             }
             final Single<StreamingHttpResponse> resp;
             if (additionalOffloads.isMetadataReceiveOffloaded() && shouldOffload.getAsBoolean()) {
-                final StreamingHttpRequest r = request;
-                resp = useExecutor.submit(
-                                () -> delegate.handle(wrappedCtx, r, responseFactory).shareContextOnSubscribe())
-                        // exec.submit() returns a Single<Single<StreamingHttpResponse>>, so flatten nested Single.
-                        // No need to apply shareContextOnSubscribe() again because unwrapped Single already shares ctx.
-                        .flatMap(identity());
+                final StreamingHttpRequest finalRequest = request;
+                // If isSendOffloaded() then we can already be offloaded by the time it subscribes to SubmitSingle.
+                // Wrap with defer to verify the shouldOffload predicate after subscribe.
+                resp = additionalOffloads.isSendOffloaded() ? Single.defer(() -> {
```

Comment:
control flow can be simplified to reduce to 1 `shareContextOnSubscribe` to make it more clear its always applied: 

```diff
diff --git a/servicetalk-http-api/src/main/java/io/servicetalk/http/api/StreamingHttpServiceToOffloadedStreamingHttpService.java b/servicetalk-http-api/src/main/java/io/servicetalk/http/api/StreamingHttpServiceToOffloadedStreamingHttpService.java
index 9d46504f8..01a8481d3 100644
--- a/servicetalk-http-api/src/main/java/io/servicetalk/http/api/StreamingHttpServiceToOffloadedStreamingHttpService.java
+++ b/servicetalk-http-api/src/main/java/io/servicetalk/http/api/StreamingHttpServiceToOffloadedStreamingHttpService.java
@@ -73,15 +73,11 @@ public class StreamingHttpServiceToOffloadedStreamingHttpService implements Stre
                 final StreamingHttpRequest finalRequest = request;
                 // If isSendOffloaded() then we can already be offloaded by the time it subscribes to SubmitSingle.
                 // Wrap with defer to verify the shouldOffload predicate after subscribe.
-                resp = additionalOffloads.isSendOffloaded() ? Single.defer(() -> {
-                    if (shouldOffload.getAsBoolean()) {
-                        return offloadHandle(useExecutor, wrappedCtx, finalRequest, responseFactory)
-                                .shareContextOnSubscribe();
-                    } else {
-                        return delegate.handle(wrappedCtx, finalRequest, responseFactory)
-                                .shareContextOnSubscribe();
-                    }
-                }) : offloadHandle(useExecutor, wrappedCtx, request, responseFactory);
+                resp = additionalOffloads.isSendOffloaded() ? Single.defer(() -> (shouldOffload.getAsBoolean() ?
+                        offloadHandle(useExecutor, wrappedCtx, finalRequest, responseFactory) :
+                        delegate.handle(wrappedCtx, finalRequest, responseFactory))
+                            .shareContextOnSubscribe()) :
+                        offloadHandle(useExecutor, wrappedCtx, request, responseFactory);
             } else {
                 resp = delegate.handle(wrappedCtx, request, responseFactory);
             }
```

Reply:
Considered that but found it less readable and easy to miss the correct bracket position to `shareContextOnSubscribe()`. 
I like ternary operator for a single case, but found it hard to read when multiple ternary operators are nested across multiple lines. Will keep it as-is.

Reply:
ack. you can also adjust parenthesis and formatting to make it more readable